### ISSUE TITLE
Reverse color palettes?

### ISSUE THREAD

**Issue by AlexanderPico** [cytoscape/py4cytoscape#62]
When using `palette_color_brewer_*` palettes like `d_RdBu` for example, the mapping is going to wrong 99% of the time, i.e., red for negative values and blue for positive. A couple suggestions:

1. reverse these palettes by default
2. offer a `reverse` Boolean parameter that gets passed down to `_palette_color_brewer`

**Comment by AlexanderPico** [comment]
In RCy3, this was necessarily implemented slightly differently (since R can't do that cool `**` thing) and there are `paletteColorBrewer*` functions that return color vectors. These are reversed by default, but can also be reversed by the user with the commonly used `rev()`function in R.

E.g., https://github.com/cytoscape/RCy3/blob/master/R/StyleMappings.R#L2539

**Comment by AlexanderPico** [comment]
Specific use case:
```
p4c.open_session()
p4c.set_node_color_mapping(**p4c.gen_node_color_map('gal80Rexp', p4c.palette_color_brewer_d_RdBu(), style_name='galFiltered Style'))
```
RAP1 has a positive expression value and GCR1 has a negative, but the typical color mapping here is reversed and I can't seem to correct it:
![Screen Shot 2021-08-14 at 11 27 25 AM](https://user-images.githubusercontent.com/1418193/129456844-4f2b4fc6-4e88-4a28-a9bc-0d93a00f1fbb.png)

**Comment by bdemchak** [comment]
Thanks, Alex … this is a nice gift.

 

I’ll work on this in the next couple of days and see if there’s something I can do that makes sense to a biologist, which is the most important standard.

 

Code-wise, no problem … but broken is broken.

**Comment by bdemchak** [comment]
Hi, Alex ...

I don't think I understand, though your example seems pretty clear.

So, GCR1 has a value of -0.53 and RAP1 has a value of 0.234 ... all as you say. Given this, I would expect GCR1 to have more of a redish color than RAP1 ... and indeed this would be the case if GCR1's color is to the left of RAP1's color on this palette. (See RdBu below)

![image](https://user-images.githubusercontent.com/3696094/129604795-77f52a16-f8b9-4f1e-a875-2f9001d12287.png)

Can you explain more about what a biologist would expect in this situation? 

I have no problem with a "reverse" parameter. All things being equal, there isn't a Cytoscape operation that duplicates this situation because Cytoscape presents only the 
qualitative colors (?) and not the Diverging colors (which contains RdBu). 

But using the same principle of lesser values being to the left of the spectrum and greater values being to the right, I tried Cytoscape out with the Paired palette (below). Indeed, Cytoscape generated colors according to principle.

![image](https://user-images.githubusercontent.com/3696094/129606061-8594681f-625a-4677-84e2-e603617c2e6f.png)

![image](https://user-images.githubusercontent.com/3696094/129606120-ecb8b6e6-307a-4c1d-a60c-9493f0376396.png)

So, I'm guessing there must be another principle that allows for applying a palette according to attribute values.

I'm guessing that it's more like ... 1) give me the palette I want, 2) let me choose whether the mapping is forward or reverse.

Can you explain more?? 

Thanks!

**Comment by AlexanderPico** [comment]
Typical divergent mapping for RdBu goes: Blue for negative, White for zero, Red for positive.  In Cytoscape GUI, you can open galFiltered and choose the RdBu divergent palette for node fill color and you'll see this by default:

![Screen Shot 2021-08-16 at 10 54 35 AM](https://user-images.githubusercontent.com/1418193/129607985-d0a3f8b3-23cb-404c-9af2-a8aea04e675b.png)

IIRC, Scooter simply reversed all the defaults for the set of divergent palettes (like I did in RCy3). In the GUI, users can manually swap the mappings along the palette glyph if they want the reverse.

**Comment by bdemchak** [comment]
Hi, Alex --

OK ... I think I see. If I look at the RCy3 StyleMappings.R code, I see that the default operation has Continuous and Sequential palettes as Brewer Colors and the Divergent palettes as the reverse of Brewer Colors.

One thing I was missing was how to apply Brewer palettes for continuous mappings in Cytoscape (i.e., double-click on the palette and choose the Current Palette selector). I had thought that palette generators happened in Cytoscape only for Discrete mappings. (This is my own fault, as Scooter has mentioned this many times and I just didn't focus on what he meant.)

So, I'll implement all of this by passing each palette generation function a reverse= parameter, which will default to True for Discrete palettes and False for Continuous and Sequential palettes. Pretty simple, and the caller can always explicitly pass reverse= if he/she cares to.

I'll let you know when I have checked this in.

I think this captures all intents in the simplest way ... please let me know if you disagree.

**Comment by AlexanderPico** [comment]
Sounds right. Happy to test when ready. 

 - Alex

**Comment by bdemchak** [comment]
Hi, Alex –

 

OK … I think I have done this …

 

Each Brewer palette function now has a reverse= parameter that defaults to False.

 

The Divergent palettes are reversed from the get-go, so they should produce the colors you’d expect. To get the original Brewer palette, the caller would use reverse=True.

 

Would you mind taking a look??

 

Thanks!

**Comment by AlexanderPico** [comment]
Trying to install dev code in notebook, but I tihnk I'm missing something... Followed this part of README:
https://github.com/cytoscape/py4cytoscape/blob/master/INSTALL.rst#install-the-development-version-python-console

But is the `git clone` step sufficient? I'm getting this error on subequent steps:

>_AttributeError: module 'py4cytoscape' has no attribute 'get_browser_client_js'_

and 

>_AttributeError: module 'py4cytoscape' has no attribute 'cytoscape_version_info'_

By the way, I'm testing the node color mapping with this tutorial run via the colab link :D
https://github.com/cytoscape/cytoscape-automation/blob/master/for-scripters/Python/basic-data-visualization.ipynb

**Comment by bdemchak** [comment]
Hmm ... I think you're executing installation instructions intended for the Python command line, not a Notebook. I don't think I've run these instructions in a while, but my first thought would be to use the instructions for Notebook use. There's a link in the Note box at the beginning of the Install section ... https://py4cytoscape.readthedocs.io/en/latest/concepts.html#jupyter-notebook ... I haven't tried these instructions in a while, either. I think it should resolve here: https://py4cytoscape.readthedocs.io/en/latest/concepts.html#jupyter-notebook

But jumping right to the example you're using, I'd say that you should execute the first cell (with the %%capture) only for a local notebook or command line Python. And you should execute the second cell (with the exec call) if you're running on a remote notebook. The two cells are mutually exclusive, either the first cell or the second cell, but not both.

In your case, there's a little more because you want to use a py4cytoscape in a Github branch. See here for the first cell you'll need: https://colab.research.google.com/github/bdemchak/cytoscape-jupyter/blob/main/gangsu/basic_protocol_2.ipynb#scrollTo=9RVNEFryzJR3 ... 

**_PY4CYTOSCAPE = 'git+https://github.com/cytoscape/py4cytoscape@0.0.10' # optional
import requests
exec(requests.get("https://raw.githubusercontent.com/cytoscape/jupyter-bridge/master/client/p4c_init.py").text)
IPython.display.Javascript(_PY4CYTOSCAPE_BROWSER_CLIENT_JS) # Start browser client**

In truth, this should work for either a local or remote notebook, thereby enabling a single development for both. 

Kozo's tutorials do this. 

A subtle point ... the documentation for 0.0.10 has this (though not well enough, I think). You can see more here: https://py4cytoscape.readthedocs.io/en/0.0.10/concepts.html#jupyter-notebook ... this will be published when 0.0.10 becomes master.

I just tried using the above snippet in the second cell of your basic viz notebook (... not executing the first cell at all), and it seems to work well.

Let's see if you can duplicate this.
